IME Events are now completely variable sized.

StayTalm opened the pull request:
IME events now have no cap.  This handles a new event that comes from native Unity that has a variable sized IME event.  Do not merge until new event is available.

-Updated IME test to still work.

StayTalm committed:
IME Events are now completely variable sized.
Updated Tests to match

StayTalm committed:
Ran Formatter

StayTalm committed:
Comments and internalize IntPtr functions

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs:
```diff
@@ -10,27 +10,64 @@ namespace UnityEngine.InputSystem.LowLevel
+            unsafe
+            {
+                int sizeInBytes = (InputEvent.kBaseEventSize + sizeof(int) + sizeof(char)) + (sizeof(char) * str.Length);
+                IntPtr evtPtr = Marshal.AllocHGlobal(sizeInBytes);
```
Use Unity NativeArray (with Allocator.Temp) and Unity Memcpy instead of Marshal.

StayTalm replied:
Not used NativeArray much, did an attempt, feel I could be missing an easier way to do it, but this passes the tests.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs:
```diff
@@ -10,27 +10,64 @@ namespace UnityEngine.InputSystem.LowLevel
+        }
+
+        /// <summary>
+        /// Queues up an IME Composition Event.!-- This event is for unit testing and debug purposes, and is slow, due to the variable way that IME Composition events work.
```
Why is it slow?

StayTalm replied:
Removed, I didn't like the AllocHGlobal and copying 1 character at a time.  Had no good justification it was slow.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs:
```diff
@@ -10,27 +10,64 @@ namespace UnityEngine.InputSystem.LowLevel
+        /// <summary>
+        /// Queues up an IME Composition Event.!-- This event is for unit testing and debug purposes, and is slow, due to the variable way that IME Composition events work.
+        /// </summary>
+        /// <param name="deviceId">The Id you'd like to send the composition event at.</param>
```
`ID of the device (see <see cref="InputDevice.deviceId") to which the composition event should be sent to. Should be an <see cref="ITextInputReceiver"/> device. Will trigger <see cref="ITextInputReceiver.OnIMECompositionChanged"/> call when processed.`

Overall, doc comments that don't really provide information and only restate what's already obvious from the code are really aggravating.

StayTalm replied:
Took that segment and did a pass on some of the other doc elements around.  Tried to add a bit of something to any of the docs, or remove the code comments entirely for things like GetEnumerator.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs:
```diff
@@ -89,50 +123,53 @@ public void Dispose()
-        public int Count => size;
+        int m_Length;
+        /// <summary>
+        /// The number of characters in the current IME Composition
```
Punctuation. Sentences in API docs should be properly formed.

StayTalm replied:
Added

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Events/IMECompositionEvent.cs:
```diff
@@ -89,50 +123,53 @@ public void Dispose()
+
+
+        /// <summary>
+        /// An Indexer into an individual character in the IME Composition. Will throw an out of range exception if the index is greater than the length of the composition.
```
Use `<exception`> tags to document exceptions.

StayTalm replied:
Done

Rene-Damm reviewed (changes requested):
No tests. Some comments RE API docs.

StayTalm committed:
PR Feedback and additional test.

StayTalm commented:
@renedamm should be good for re-review, moved to using Unity's unsafe & native array tools, improved the docs.  

Added 1 test, and was already using the existing IME event test to verify that the composition events get received properly.

StayTalm committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into Variable-Sized-IME

StayTalm commented:
@Rene-Damm  
Looks like NativeArray is missing Reinterpret Store on 2019.1, causing it to fail to compile.  
Should I revert, is there a better way to use NativeArray, or should I start testing with later versions?

StayTalm committed:
Second attempt using NativeArray

StayTalm committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into Variable-Sized-IME

StayTalm committed:
Run Formatter

StayTalm committed:
Swap back to keeping the exisiting event format and creating a second method for getting hte NativeArray to avoid breaking API.

StayTalm committed:
Formatting

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1120) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1120) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/InputSystem?pullRequest=1120) it.</sub>

jimon commented:
I'm recreating this work in #1628
